extra_data/components.py

```diff
@@ -926,6 +738,311 @@ def zip_trains_pulses(trains, pulses):
+        split = self.split_trains(frames_per_part=min(1024 ** 3 / entry_size, 256))
+
+        arr = concatenate([from_delayed(
+            delayed(c.ndarray)(fill_value=fill_value, astype=astype),
+            shape=c.shape, dtype=self.dtype
+        ) for c in split], axis=1)
+
+        if labelled:
+            from xarray import DataArray
+            coords = {'module': self.modules, 'trainId': self.train_id_coordinates()}
+            return DataArray(arr, dims=self.dimensions, coords=coords)
+
+        return arr
+
+
+class XtdfImageMultimodKeyData(MultimodKeyData):
+    _sel_frames_cached = None
+
+    def __init__(self, det: XtdfDetectorBase, key, pulse_sel=by_index[0:MAX_PULSES:1]):
+        super().__init__(det, key)
+        self.det = det  # Makes PyCharm happy that det is XtdfDetectorBase
+        self._pulse_sel = pulse_sel
+        entry_shape = self._eg_keydata.entry_shape
+        self._extraneous_dim = (len(entry_shape) >= 1) and (entry_shape[0] == 1)
+
+    @property
+    def ndim(self):
+        return super().ndim - (1 if self._extraneous_dim else 0)
+
+    def _all_pulses(self):
+        psv = self._pulse_sel.value
+        return isinstance(psv, slice) and psv == slice(0, MAX_PULSES, 1)
+
+    def _shape(self, module_gaps=False, roi=()):
+        module_dim = self.det.n_modules if module_gaps else len(self.modno_to_keydata)
+        nframes_sel = len(self.train_id_coordinates())
+
+        entry_shape = self._eg_keydata.entry_shape
+        if self._extraneous_dim:
+            entry_shape = entry_shape[1:]
+
+        return (module_dim, nframes_sel) + roi_shape(entry_shape, roi)
+
+    @property
+    def shape(self):
+        return self._shape()
+
+    def train_id_coordinates(self):
+        # XTDF 'image' group can have >1 entry per train
+        a = self.det.train_ids_perframe
+        # Only allocate sel_frames array if we need it:
+        if not self._all_pulses():
+            a = a[self._sel_frames]
+        return a
+
+    @property
+    def dimensions(self):
+        ndim_inner = self.ndim - 2
+        # TODO: this assumes we can tell what the axes are just from the
+        # number of dimensions. Works for the data we've seen, but we
+        # should look for a more reliable way.
+        if ndim_inner == 3:
+            # image.data in raw data
+            entry_dims = ['data_gain', 'slow_scan', 'fast_scan']
+        elif ndim_inner == 2:
+            # image.data, image.gain, image.mask in calibrated data
+            entry_dims = ['slow_scan', 'fast_scan']
+        else:
+            # Everything else seems to be 1D, but just in case
+            entry_dims = [f'dim_{i}' for i in range(ndim_inner)]
+        return ['module', 'train_pulse'] + entry_dims
+
+    # Used for .select_trains() and .split_trains()
+    def _with_selected_det(self, det_selected):
+        return XtdfImageMultimodKeyData(det_selected, self.key, self._pulse_sel)
```

Comment:
Do you think it might be beneficial to use `self.__class__` instead to facilitate further inheritance?
(same for method below)

Reply:
Theoretically, maybe, but I'd strongly discourage anyone subclassing this outside of EXtra-data - it would be very brittle - and if we subclass it internally, we can fix that when we come to it.